app/drivers/zephyr/dts/bindings/kscan/zmk,kscan-gpio-matrix.yaml

```diff
@@ -8,6 +8,9 @@ compatible: "zmk,kscan-gpio-matrix"
 include: kscan.yaml
 
 properties:
+  ports:
```

Comment:
I agree that we should be able to just do this automatically from the existing data. I think it could just replace the existing code rather than being an option you have to enable, since reading by port is always going to be more efficient than reading pin by pin (gpio_pin_get_dt just calls gpio_port_get and then masks the result for the given pin).

It doesn't necessarily need to be done with the C preprocessor though (I can't think of a way to do that either). It could instead be built dynamically from `inputs` during the init function. Ideally this code could be shared between matrix and direct GPIO drivers too.

Not the simplest thing in the world, but it might look something like this:
```c
// kscan/gpio.h
struct kscan_gpio_list {
    const struct gpio_dt_spec* gpios;
    size_t len;
};

struct kscan_gpio_port_list {
    struct device** ports;
    size_t len;
};

/**
 * Gets the set of ports containing the given GPIOs.
 */
struct kscan_gpio_port_list kscan_gpio_get_ports(const struct kscan_gpio_list* gpios);

static inline bool kscan_gpio_is_on_port(const struct gpio_dt_spec* gpio, const device* port) {
    return gpio->port == port;
}

/**
 * Given the result of gpio_port_get() for a GPIO's port, gets whether that GPIO is active.
 */
static inline bool kscan_gpio_get_from_port(const struct gpio_dt_spec* gpio, gpio_port_value_t value) {
    return (value & BIT(gpio->pin)) != 0;
}
 
// kscan/gpio.c
static bool contains_port(const kscan_gpio_port_list* list, const struct gpio_dt_spec* gpio) {
    for (int i = 0; i < list->len; i++) {
        if (list->ports[j] == gpio->port) {
            return true;
        }
    }
    return false;
}

static void shrink_to_fit(kscan_gpio_port_list* list) {
    struct device** original = list->ports;
    list->ports = k_calloc(list->len, sizeof(struct device*));
    
    for (int i = 0; i < list->len; i++) {
        list->ports[i] = original[i];
    }
    
    k_free(original);
}

struct kscan_gpio_port_list kscan_gpio_get_ports(const struct kscan_gpio_list* gpios) {
    struct kscan_gpio_ports_list result = {
        .ports = k_calloc(gpios->len, sizeof(struct device*)),
        .len = 0,
    };
    
    for (int i = 0; i < gpios->len; i++) {
        if (!contains_port(&result, &gpios->gpios[i]) {
            result.ports[result.len] = gpios->gpios[i].port;
            result.len++;
        }
    }
    
    shrink_to_fit(&result);
    return result;
}

// kscan/kscan_gpio_matrix.c
...
for (int p = 0; p < config->ports.len; p++) {
    const struct device* in_port = config->ports.ports[p];
    gpio_port_value_t port_value;
    int err = gpio_port_get(in_port , &port_value);
    if (err) {
        LOG_ERR("Failed to read port %s: %i", in_port->name, err);
        return err;
    }
     
    for (int i = 0; i < config->inputs.len; i++) {
        const struct gpio_dt_spec* in_gpio = &config->inputs.gpios[i];
        if (!kscan_gpio_is_on_port(in_gpio, in_port)) {
            continue;
        }
        const int index = state_index_io(config, i, o);
        const bool active = kscan_gpio_get_from_port(in_gpio, port_value);
        ...
```

Reply:
If we create the port list dynamically in the init function why not use `sys_slist_t` where each container holds the port and a kscan_gpio_list of the used pins of this port?

So the read loop could call the FOR_EACH macro and iterate over ports and given port pins.

Reply:
You could do that, though it would use a bit more memory than this, and you'd still have to dynamically sort the inputs into kscan_gpio_lists, so I doubt the code would be any simpler. It could be more efficient since you wouldn't need the `kscan_gpio_is_on_port()` check, but I expect the improvement would be pretty negligible since you probably only have 1-4 ports used, and looping through an array and checking if pointers match is quick.

I don't really have any strong feelings either way.

Reply:
The idea to include pins within the ports will not work because `state_index_io` would not get the correct index.

I updated the implementation to use `sys_slist_t` because, in my opinion, it is cleaner instead of using multiple functions to do what a list already can do.

At the moment, I did not create an additional file (gpio.h/c) because I would like to ask if shared code between direct and matrix, like `struct kscan_gpio_list` and `#define KSCAN_GPIO_LIST(gpio_array)`, should also be moved to this file?

Reply:
Big note: any solution should really not depend on dynamic allocation. We currently depend on that for our eventing system and the optional display bits, but I have ideas for fixing eventing to avoid that. As much as possible, we should be doing static allocation, especially at the lower level like drivers.

Reply:
I do not think it is possible to generate a unique port list without dynamic memory allocation in the init function.

Another way to read only ports and not pins without the usage of such a port list would be to iterate somewhat smarter over the input pin list.

In the worst case, each input pin has another port and if the `input_updated_mask` is `uint64_t`, the input length is limited to 64 pins.
The limitation could be omitted if an array is used instead of bitmasking.

```
uint64_t input_updated_mask = 0;

// always begin at the port of the first input pin
const struct device *input_port = config->inputs.gpios[0].port;

for (int i = 0; i < config->inputs.len; i++)
        {
            gpio_port_value_t port_value;
            
            // fetch port state
            int err = gpio_port_get(input_port, &port_value);
            if (err)
            {
                LOG_ERR("Failed to read port %s: %i", input_port->name, err);
                return err;
            }

            // iterate over each pin, start at the current pin
            for (int j = i; j < config->inputs.len; j++)
            {
                // update each pin that has the correct port
                if (input_port == config->inputs.gpios[j].port)
                {
                    const int index = state_index_io(config, j, o);
                    const bool active = (bool)(port_value & BIT(config->inputs.gpios[j].pin));

                    debounce_update(&data->matrix_state[index], active,
                                    config->debounce_scan_period_ms, &config->debounce_config);

                    // mark pin as updated
                    input_upated_mask |= BIT(j);
                }
            }

            input_port = NULL;
            
            // search for the next pin that is not updated -> port that is not updated
            for (int j = i; j < config->inputs.len; j++) {
                if (!(bool)(input_updated_mask & BIT(j))) {
                    input_port = config->inputs.gpios[j].port;

                    // skip to the pin of the new found port
                    i = j - 1;
                }
            }

            // if no port is found end the loop
            if (input_port == NULL) {
                break;
            }
        }
```

Reply:
I believe I've found a simple way to do this without any dynamic allocation. I've gotten it compiling but haven't yet tested it.

If you simply sort the list of inputs by port, then as you iterate through the list, you can do a port read any time the current element's port doesn't match the previous one's. This does mean you have to track each pin's original index in the list though.

Reply:
My attempt at doing this without any dynamic allocations is up at #1571.